### DISCUSSION TITLE
Set company-idle-delay with a function.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [company-mode/company-mode#1379] [Q&A]
I try to set company-idle-delay with a function as follows:
```emacs
;; Method 1
   company-idle-delay (lambda ()
			(cond
			 ((derived-mode-p 'python-mode) 0.2)
			 ((derived-mode-p 'latex-mode) 0.2)
			 (t 0.5)
			 ))

;; Method 2
   company-idle-delay
   (if (memq major-mode '(python-mode latex-mode))
       0.2
     0.5)

;; Method 3
   company-idle-delay
   (if (or (derived-mode-p 'python-mode)
           (derived-mode-p 'latex-mode))
       0.2
     0.5)

;; Method 4
   ;; this method doesn't work exactly as others:
   company-idle-delay
   (if (memq (derived-mode major-mode) '(python-mode latex-mode))
       0.2
     0.5)
```
It turns out that the completion behavior of the last method isn't exactly the same as others. But I can't see what is subtle differences among these methods. Any hints will be appreciated.

Regards,
Zhao

**Comment by hongyi-zhao** [comment]
It seems that `lambda` method must be used here. So, all the following methods are equivalent:
```emacs
   company-idle-delay (lambda ()
			(cond
			 ((derived-mode-p 'python-mode) 0.2)
			 ((derived-mode-p 'latex-mode) 0.2)
			 (t 0.5)
			 ))

   company-idle-delay
   (lambda ()
     (cond ((memq major-mode '(python-mode latex-mode)) 0.2)
           (t 0.5)))

   company-idle-delay
   (lambda ()
     (if (memq major-mode '(python-mode latex-mode))
	 0.2
       0.5)
     )

   company-idle-delay
   (lambda ()
     (if (or (eq major-mode 'python-mode)
             (eq major-mode 'latex-mode))
	 0.2
       0.5))
```

The above methods are all different ways of defining the `company-idle-delay` function for the Emacs editor. 

The first method uses the `cond` function to check if the current mode is either Python or LaTeX, and sets the delay to 0.2 seconds in those cases. Otherwise, it sets the delay to 0.5 seconds.

The second method uses the `memq` function to check if the major mode is either Python or LaTeX, and sets the delay to 0.2 seconds in those cases. Otherwise, it sets the delay to 0.5 seconds.

The third method uses an `if` statement to check if the major mode is either Python or LaTeX, and sets the delay to 0.2 seconds in those cases. Otherwise, it sets the delay to 0.5 seconds.

The fourth method is similar to the second method, but uses `eq` instead of `memq` to check for equality.

Overall, all of these methods achieve the same goal, but use slightly different syntax and functions to do so. The choice of which method to use may depend on personal preference or the specific requirements of the project.

**Comment by dgutov** [comment]
It depends on when you do these assignments (does `major-mode` have the expected value yet?), and whether you set buffer-local values (you should, e.g. with `setq-local`).

That is, those questions are important for methods 2-4.

Method 1 does a "late binding" or sorts, so it answers the question 1 as "yes" automatically, and it doesn't need a buffer-local value.

**Reply by hongyi-zhao** [reply]
Thank you for your comments and explanations. Here are some of my additional thoughts on the topic:

In short: The timing of these assignments and whether they are buffer-local do play an important role in determining which method to use.

Method 1 uses a "late binding" approach, which means that it will automatically have the correct value of `major-mode` at the time it is evaluated. This method does not require a buffer-local value because it is evaluated each time `company-idle-delay` is called.

Methods 2-4 need to be evaluated at the time the buffer is opened, or the mode is changed, so they require buffer-local values to ensure that the correct value of `major-mode` is used. Method 2 is similar to method 1, but it uses `memq` instead of `derived-mode-p` to check the value of `major-mode`. Method 3 uses an `if` statement to check whether `major-mode` is in a list of modes, and then sets `company-idle-delay` accordingly. Method 4 uses `eq` to check whether `major-mode` is equal to specific modes and then sets `company-idle-delay` accordingly.

Overall, the choice of method depends on the specific use case and the desired behavior of `company-idle-delay`. If one need to dynamically adjust the value of `company-idle-delay` based on the current major mode, then methods 1 or 2 might be appropriate. If one need to set a specific value of `company-idle-delay` for certain modes, then methods 3 or 4 might be more appropriate. And as you mentioned, setting buffer-local values with `setq-local` is important to ensure that the correct value of `company-idle-delay` is used in each buffer.

Any further comments/corrections will be appreciated.

Regards,
Zhao

**Reply by yugaego** [reply]
>  If one need to dynamically adjust the value of company-idle-delay based on the current major mode, then methods 1 or 2 might be appropriate.

I'm tempted to ask: why not to simply follow --a more idiomatic, some might argue-- approach using a [mode hook and setq-local](https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#configure-different-backends-for-different-modes)?
Do you have some specific requirements maybe or it's a matter of preference, investigation?

**Reply by hongyi-zhao** [reply]
With the method mentioned above, you need to create separate [mode hook and setq-local](https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#configure-different-backends-for-different-modes)'s for the corresponding modes/packages. At least from my point of view, it seems to be a bit cumbersome than the first method discussed here.

**Reply by yugaego** [reply]
Gotcha, thanks.
